### DISCUSSION TITLE
Optimal Configuration for parent scaling to 100s of children (short retention time) : RAM Issue

### DISCUSSION THREAD

**Discussion by AidanMcP** [netdata/netdata#15225] [Q&A]
Hi Netdata community – I would really appreciate your guidance and expertise on the following.

I have:
- 100s of Netdata children nodes streaming to a Netdata parent. Typically 150+ - could grow towards 1K
- Each node has potential to produce several 100 metrics per second, so could have 10K+ metrics per second
- The consumer (Prometheus) scrapes every 5seconds, so Netdata needs only to store for a short time circa 5-10 seconds
- Given the transient nature of the metrics, I had opted for memory mode = “ram” but suspect that is wrong based on the links below:
- am seeing unbounded memory growth with netdata-1.34.1 such that OOM killer interviews – seems like a leak, TBD
- is a container based install, k8s, redhat 8, with no specific kernel, OS/KSM optimizations for Netdata thus far.

Ref:
https://learn.netdata.cloud/docs/configuring/optimizing-metrics-database/database-modes-for-parent-child-setups and
https://community.netdata.cloud/t/insane-netdata-memory-usage/3342/2 (from Christopher_Akritid1)

My questions are:

1) Even with very transient very short-lived metrics, from 100+ connections – is dbengine the only mode to use on Parent and preferable to “ram”?
2) Are there other optimizations for this scenario ? e.g. How can I force Netdata to only store a few seconds of metrics.?
3) Is there a known memory leak in netdata-1.34.1 and/or a reason to move to netdata-1.39.1 ?
Note: Tag names and format seem to have changed in 1.39.1 so am hesitant to upgrade unless there’s a compelling reason.
4) Are there any recommended tools, frameworks or techniques to simulate 1000’s of metrics being streamed to Parent Netdata.
This would greatly help simulate and verify setups in a friendly environment.
5) Assuming an optimal setup, are there sizing rules to estimate RAM usage per metric.

Help with any of these questions would be greatly appreciated. 
Thanks,
Aidan

```
[global]
    #Explicitly setting netdata agent name
    hostname = <mymachine>

    run as user = netdata

    #memory mode defined as save
    memory mode = ram

    # history controls the size of the database in RAM, except for memory mode = dbengine
    history = 30

    process nice level = 0
    OOM score = -1000

    # Expose the netdata port to all
    bind to = localhost *

>> [edited for brevity - disabled unrequired plug-ins]

[health]
        enabled = no

[ml]
        enabled = no

[host labels]
        Region=
        Node_Type=<mytype>

```

**Comment by andrewm4894** [comment]
Hi @AidanMcP interesting set up and questions - am asking the agent team for some thoughts.

**Reply by mjksohonet** [reply]
Hi @AidanMcP 
if you have some spare h/w, you might like to install the free version of XCP-ng 8.2.
This comes with netdata 1.19.
One of our test servers has 12 VMs and netdata is collecting 3400+ metrics

**Comment by MrZammler** [comment]
Hi @AidanMcP 

I don't remember a memory leak in 1.34.1, but that's not to say that there isn't one. 1.39.1 or 1.40.0 should fare better in the ram department, but really depends on the specific install.

Also checking your `OOM score = -1000` setting, likely it doesn't work, our default is 0, which should make it less likely to be killed by OOM.

**Comment by vkalintiris** [comment]
@AidanMcP Not sure if this helps, but I opened https://github.com/netdata/netdata/pull/13962 which should make experimentation a little bit easier and faster. In my tests, I generated 3200 charts with 5  dimensions each, and I backfilled 7 days of data with `update every = 1` second. It took a little bit less than 2 hours, on my system, for the plugin to finish the backfilling. The agent ended up using ~1.4 GiB.

Once it's merged, you could experiment with a different number of charts/dimensions and backfill/update-every durations to get an understanding of how resource consumption is affected. Note, that the plugin attaches all (randomly-generated) data to localhost, which means that in practice you'll get worse performance results than what you would get by using real-children agents streaming to a parent.
